docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,315 @@
+
+1.  Storage classes should be able to hold a set of provisioner parameters
+2.  The provisioner should be passed the claim and the details of the storage
+    class it is provisioning for
+
+#### Out-of-tree provisioners
+
+One of our goals is to enable administrators to create out-of-tree
+provisioners, that is, provisioners whose code does not live in the Kubernetes
+project.  Our experience since the 1.2 release with dynamic provisioning has
+shown that it is impossible to anticipate every aspect and manner of
+provisioning that administrators will want to perform.  The proposed design
+should not prevent future work to allow out-of-tree provisioners.
+
+## Analysis
+
+### Summary of alpha functionality
+
+The alpha functionality for dynamic provisioning works as follows:
+
+1.  A cluster has only a single provisioner at a time, dictated by the cloud
+    provider
+2.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.  If a suitable volume is found, the
+    controller binds the claim and volume together and carries on normally.
+3.  If no volume is found for the claim, and the claim has the annotation
+    `volume.alpha.kubernetes.io/storage-class`, the provisioner is invoked
+    to provision a volume to fill that claim
+4.  All provisioners are in-tree; they implement an interface called
+    `ProvisionableVolumePlugin` which has a method, `NewProvisioner()`,
+    that returns a new `Provisioner` instance
+5.  The controller calls the provisioner's `Provision()` method; `Provision`
+    is responsible for provisioning the volume in the cloud provider and
+    returns an `api.PersistentVolume` and an error
+6.  If `Provision` returns an error, the controller stops trying to provision
+7.  If `Provision` returns no error, the controller creates the returned
+    `api.PersistentVolume`, already bound to the claim
+  1.  If the create operation fails, it is retried
+  2.  If the create operation never succeeds, the controller attempts to delete
+      the provisioned volume and creates an event on the claim
+
+### When should we provision volumes?
+
+It is sensible to provision volumes only when an existing volume cannot
+satisfy a claim.  This ensures that existing, already allocated resources are
+used before additional resources are allocated.  In this regard, the existing
+logic need not change much when storage classes are added.  The primary
+difference is that in the existing design, there is no way to parameterize
+provisioners or have more than one provisioner per cluster.  Therefore, once
+we know that we have to provision a volume to satisfy a claim, we need to
+determine which storage class to use for the claim.
+
+### How should a claim be matched to a storage class?
+
+We've discussed so far that volume claims and storage classes will be matched
+using the claim's label selector.  With that said, there are a number of small
```

Comment:
I ACK the idea that the user shouldn't need to know the difference.  Do we really think that mixing static and dynamic is going to be common?

The part I dislike here is that selecting by labels feels like a very roundabout way of doing things.  Classes have a `metadata.name` which is NOT a label.  I don't think the user should be saying "what combination of labels do I need to apply to choose precisely the 'triangle' class, deterministically".  They should just say "triangle".

What about promoting a particular label when applying to classes?  If the user files a PVC with selector `{"volume.kubernetes.io/class": "triangle", "failure-domain.beta.kubernetes.io/region": "us-west"}`, we first try to find an extant volume that matches that (again have to answer what "matches" means).  If that fails, then we try to provision, but the ONLY class that is considered is "triangle".  The provisioner MUST comprehend and satisfy all additional labels or fail (this is not documented in here AFAICT, and it's important).  The resulting volume has the same labels as the claim's selector.

This dodges the "what if multiple provisioners match" mess, and makes the class name meaningful and consistent.  But it feels like an abuse of labels, which are not supposed to be semantically meaningful to the system.

The risk of just being labels is that cluster-admin for A decides to label his classes with colors, and cluster-admin for B labels with shapes, and any user who has to go between them has to pay attention to the difference.  Maybe that's OK, and I am over-focusing on consistency?  Maybe a concrete and non-trivial example of how we expect this to be used (people will copy what we use as examples)?